{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python\n",
    "# coding: utf-8\n",
    "\n",
    "# In[38]:\n",
    "\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re \n",
    "import matplotlib.pyplot as plt \n",
    "import sys\n",
    "import time\n",
    "\n",
    "\n",
    "\n",
    "#Input file must be given to run the code.\n",
    "#For Example Run(in terminal) 'python3 Bioinformatics input.xlsx'\n",
    "\n",
    "#Filename =sys.argv[1] # Read input file \n",
    "df = pd.read_excel('updatedBigDataInput.xlsx')# Read EXCEL Files and storing in database 'df'\n",
    "\n",
    "\n",
    "data= pd.DataFrame()# Creating New Data Frame \n",
    "\n",
    "\n",
    "# Storing Usefull information From the database 'df' \n",
    "data['Entry']=df['Entry'] #Storing Entry\n",
    "data['Entry name']=df['Entry name']  #Storing Entry Name\n",
    "data['Protein names']=df['Protein names']  #Storing Protein Names\n",
    "data['Gene names']=df['Gene names'] # Storing Gene Names\n",
    "data['Length']=df['Length'] # Storing Length\n",
    "data['Organism']=df['Organism'] # Storing Type of Organism\n",
    "data['Mass']=df['Mass'] #Storing Mass\n",
    "data['Status']=df['Status']\n",
    "\n",
    "#Now removing the extra data we dont need and containg only organism 'Human'\n",
    "human_data=data[data['Organism'].str.contains(\"Homo sapiens\")] #Selecting only organism which contains 'Homo sapiens'\n",
    "data= human_data\n",
    "\n",
    "data['Disulfide bond']=df['Disulfide bond'] # Storing Disulfide bond \n",
    "data['Glycosylation']= df['Glycosylation']# Storing Glycosylation Position\n",
    "data=data.dropna().reset_index(drop=True) # Drop all rows which contains Not a number and reset the index \n",
    "\n",
    "# Working on getting the relative positions of the Disulfide bond\n",
    "disulfide_column1= data['Disulfide bond'] # Storing data temporary as 'disulfide_column'\n",
    "disulfide_column2= data['Disulfide bond']\n",
    "\n",
    "def get_sulfide_value(newdata):\n",
    "    return re.findall('\\d+\\..\\d+',newdata)\n",
    "\n",
    "bond= disulfide_column1.apply(get_sulfide_value) # Function call which gives all the positons \n",
    "data['Disulfide bond']=disulfide_column2.apply(get_sulfide_value)# Storing extracted disulfide value in data Frame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_int_value(data):\n",
    "    return re.findall('\\d+',data)\n",
    "def get_float_value(data):\n",
    "    return re.findall('\\d+\\.\\d+',data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "Glyco_data= data['Glycosylation'] # Storing data temporary as 'Glyco_data'\n",
    "def getGlycoNLinked(data):   #Making a Function to get the positions\n",
    "    return re.findall('CARBOHYD \\d+',data) # The return data will contain 'CARBOHYD position #'\n",
    "data['Glycosylation']= Glyco_data.apply(getGlycoNLinked) # the return data will be like 'CARBOHYD 110'\n",
    "#Now Removing the extra word 'N-linked' and getting all positions of Glycosylation.\n",
    "temp_data= data['Glycosylation'].astype(str)\n",
    "data['Glycosylation']= temp_data.apply(get_int_value) \n",
    "\n",
    "\n",
    "# Creating a temporary dataframe for total length in columns and length in rows(like number of pairs in each index).\n",
    "a= pd.DataFrame()\n",
    "b= pd.DataFrame()\n",
    "a =pd.Series(data['Glycosylation'])\n",
    "b =pd.Series(data['Disulfide bond'])\n",
    "lenB=b.str.len()\n",
    "lenA=a.str.len()\n",
    "\n",
    "rel_pos = pd.Series(a.size,dtype=np.str)\n",
    "for i in range(a.size):\n",
    "    rel_pos[i]=''\n",
    "    \n",
    "InsidePairs=pd.Series(b.size,dtype=np.str)\n",
    "InsidePairsLength=np.zeros(b.size,dtype=np.int)\n",
    "for i in range(a.size):\n",
    "    InsidePairs[i]=''\n",
    "\n",
    "# Now using loop and properties of Python library 're' to get values and finding positions between disulphide bond and glycosation.\n",
    "i=0\n",
    "while i<b.size:\n",
    "    for j in range(lenA[i]):\n",
    "        Position=int(a[i][j])\n",
    "        rel_pos[i]=rel_pos[i]+str(a[i][j])+'{|'\n",
    "        for k in range(lenB[i]):\n",
    "            split = re.findall('\\d+',b[i][k]) \n",
    "            firstPair= int(split[0])\n",
    "            secondPair=int(split[1])\n",
    "            if Position< firstPair:\n",
    "                rel_pos[i] = rel_pos[i]+'o'+str(Position-firstPair)\n",
    "                rel_pos[i] = rel_pos[i]+'o'+str(Position-secondPair)\n",
    "            elif Position>= firstPair and Position<=secondPair:\n",
    "                rel_pos[i]=  rel_pos[i]+'i' + str(Position-firstPair)\n",
    "                rel_pos[i] = rel_pos[i]+'i'+ str(Position-secondPair)\n",
    "                InsidePairs[i]= InsidePairs[i] + str('('+ str(firstPair)+','+str(secondPair)+')')\n",
    "                InsidePairsLength[i]=InsidePairsLength[i]+ (secondPair-firstPair)\n",
    "            else:\n",
    "                rel_pos[i] = rel_pos[i]+'o'+str(Position-firstPair)\n",
    "                rel_pos[i] = rel_pos[i]+'o'+str(Position-secondPair)\n",
    "            rel_pos[i]=rel_pos[i]+'|'    \n",
    "        rel_pos[i]=rel_pos[i]+'}'\n",
    "    i=i+1\n",
    "data['rel_pos']= rel_pos\n",
    "data['In_Gly_Dis_Pair']=InsidePairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Caluclating Score of each pairs of disulfide bonds\n",
    "# O score if other pair outside.\n",
    "# 0.5 score if other pair half inside and half outside.\n",
    "# 1.0 score if other pair full inside.\n",
    "array = pd.Series(a.size,dtype=np.str)\n",
    "for x in range(a.size):\n",
    "    array[x]=''\n",
    "\n",
    "for i in range(0,len(data)):\n",
    "    temp_array = np.zeros(lenB[i])\n",
    "    for j in range(0,lenB[i]):\n",
    "        for k in range(j+1,lenB[i]):\n",
    "            if k!=lenB[i]:\n",
    "                split1 = re.findall('\\d+',data['Disulfide bond'][i][j]) \n",
    "                p1e1=int(split1[0])\n",
    "                p1e2=int(split1[1])\n",
    "                split2 = re.findall('\\d+',data['Disulfide bond'][i][k]) \n",
    "                p2e1=int(split2[0])\n",
    "                p2e2=int(split2[1])\n",
    "            \n",
    "                if(p2e1>p1e2):\n",
    "                    temp_array[j]=temp_array[j]+0\n",
    "                    temp_array[k]=temp_array[k]+0\n",
    "                elif (p2e1<p1e2) and (p2e2>p1e2):\n",
    "                    temp_array[j]=temp_array[j]+0.5\n",
    "                    temp_array[k]=temp_array[k]+0.5    \n",
    "                elif (p2e1<p1e2) and (p2e2<p1e2):\n",
    "                    temp_array[j]=temp_array[j]+1.0\n",
    "                    temp_array[k]=temp_array[k]+0.0\n",
    "        array[i]=array[i]+str(temp_array[j])+','\n",
    "data['Disulfide_score']= array.apply(get_float_value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#Claculating disulfide pair score average \n",
    "disulfie_pair_score_average = np.zeros(len(data),dtype=np.float)\n",
    "for i in range(len(data)):\n",
    "    for j in range(len(data['Disulfide_score'][i])):\n",
    "        disulfie_pair_score_average[i]= disulfie_pair_score_average[i]+ float(data['Disulfide_score'][i][j])\n",
    "data['disulfie_pair_score_average']=disulfie_pair_score_average"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:72: RuntimeWarning: invalid value encountered in long_scalars\n"
     ]
    }
   ],
   "source": [
    "## Calculating Interbond Distance\n",
    "interbond_distance= pd.Series(a.size,dtype= np.str)\n",
    "\n",
    "for i in range(a.size):\n",
    "    interbond_distance[i]=''\n",
    "\n",
    "for i in range(a.size):\n",
    "    \n",
    "    if lenB[i]==0:\n",
    "        interbond_distance[i]='No pair'\n",
    "        \n",
    "    if lenB[i]==1:\n",
    "        interbond_distance[i]='Only One pair'\n",
    "    \n",
    "    for r in range(lenB[i]-1):\n",
    "        split1= re.findall('\\d+',data['Disulfide bond'][i][r])\n",
    "        split2= re.findall('\\d+',data['Disulfide bond'][i][r+1])\n",
    "        interbond_distance[i]= interbond_distance[i]+ str(int(split2[0])-int(split1[1]))+ '|'\n",
    "        \n",
    "data['interbond_distance']=interbond_distance\n",
    "\n",
    "#Intrabond distance formation\n",
    "intrabond=pd.Series(b.size,dtype=np.str)\n",
    "for i in range(a.size):\n",
    "    intrabond[i]=''\n",
    "j=0;\n",
    "while j<b.size:\n",
    "    for k in range(lenB[j]):\n",
    "        split = re.findall('\\d+',b[j][k]) \n",
    "        firstPair= int(split[0])\n",
    "        secondPair=int(split[1])\n",
    "        intrabond[j]=intrabond[j]+str(secondPair-firstPair)+'|'    \n",
    "    j=j+1\n",
    "data['intrabond']=intrabond\n",
    "\n",
    "\n",
    "#number of pairs of disuphide bonds\n",
    "data['total sulphide bonds']=lenB\n",
    "\n",
    "#number of glcosylation positions\n",
    "data['total glyco positions']=lenA\n",
    "\n",
    "\n",
    "#1st disulphide pair to N-terminus distance\n",
    "first_pos = pd.Series(a.size,dtype=np.int)\n",
    "for x in range(a.size):\n",
    "    if(data['Disulfide bond'][x]!=[]):\n",
    "        split = re.findall('\\d+',data['Disulfide bond'][x][0]) \n",
    "        firstPair= int(split[0])\n",
    "        first_pos[x]= firstPair\n",
    "data['first_position_occurance']=first_pos\n",
    "\n",
    "\n",
    "#last disulphide pair to C-terminus distance\n",
    "last_pos= pd.Series(a.size,dtype=np.int)\n",
    "for x in range(a.size):\n",
    "    if(data['Disulfide bond'][x]!=[]):\n",
    "        split= re.findall('\\d+',data['Disulfide bond'][x][(lenB[x]-1)])\n",
    "        lastPair= int(split[1])\n",
    "        last_pos[x] = lastPair \n",
    "data['last_position_occurance']=last_pos\n",
    "\n",
    "#Calculate average distance of intrabond distance\n",
    "average= data['intrabond'].apply(get_int_value)\n",
    "lenIntra=average.str.len()\n",
    "\n",
    "averageintrabond= np.zeros(a.size)\n",
    "for i in range(a.size):\n",
    "    totalsum=0\n",
    "    for j in range(lenIntra[i]):\n",
    "        totalsum= totalsum+int(average[i][j])\n",
    "    averageintrabond[i]=totalsum/lenIntra[i]\n",
    "data['average_Intrabond']=averageintrabond\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#calculating average distance of those pairs of disulphide bonds that have Glycosylation Inside.\n",
    "temp_length=pd.Series(b.size,dtype=np.str)\n",
    "#for i in range(a.size):\n",
    "#    temp_length[i]=''\n",
    "inside_length = np.zeros(a.size,dtype=np.int)\n",
    "for r in range(a.size):\n",
    "    inside_length[r]=len(re.findall('\\d+',data['In_Gly_Dis_Pair'][r]))/2\n",
    "    temp_length[r]=re.findall('\\d+',data['In_Gly_Dis_Pair'][r])\n",
    "    if inside_length[r]!=0:\n",
    "        InsidePairsLength[r]=InsidePairsLength[r]/inside_length[r]\n",
    "data['avg_Inside_Pairs_Length']=InsidePairsLength"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#If the glyco is inside then putting the sulphide position 'nil' at that point.\n",
    "i=0\n",
    "while i<b.size:\n",
    "    for j in range(lenA[i]):\n",
    "        Position=int(a[i][j])\n",
    "        for k in range(lenB[i]):\n",
    "            if bond[i][k]!='nil':\n",
    "                split = re.findall('\\d+',b[i][k]) \n",
    "                firstPair= int(split[0])\n",
    "                secondPair=int(split[1])\n",
    "                if Position>= firstPair and Position<=secondPair:\n",
    "                    bond[i][k]='nil'\n",
    "    i=i+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Calculating length between sulfide bonds.\n",
    "temp=pd.Series(b.size,dtype=np.str)\n",
    "for i in range(a.size):\n",
    "    temp[i]=''\n",
    "\n",
    "for i in range(len(bond)):\n",
    "    for j in range(lenB[i]):\n",
    "        if bond[i][j]!='nil':\n",
    "            split= re.findall('\\d+',bond[i][j])\n",
    "            temp[i]=str(temp[i] + str(int(split[1])-int(split[0]))+'|')\n",
    "data['glyco_outside_bond']=bond"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get the integer values.\n",
    "intrabond_outside= temp.apply(get_int_value)\n",
    "data['intrabond_glyco_outside_bond']=intrabond_outside"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#calculating avegrage lenth of sulphide bonds which do not have glyco inside.\n",
    "\n",
    "new =pd.Series(intrabond_outside)\n",
    "newlen=new.str.len()\n",
    "average= np.zeros(a.size)\n",
    "for i in range(a.size):\n",
    "    totalsum=0\n",
    "    for j in range(newlen[i]):\n",
    "        totalsum= totalsum+int(intrabond_outside[i][j])\n",
    "    if newlen[i]!=0:\n",
    "        average[i]=totalsum/newlen[i]\n",
    "data['average_Intrabond_outside_glyco_bond']=average"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "data.to_excel('output.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry name</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene names</th>\n",
       "      <th>Length</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Mass</th>\n",
       "      <th>Status</th>\n",
       "      <th>Disulfide bond</th>\n",
       "      <th>Glycosylation</th>\n",
       "      <th>...</th>\n",
       "      <th>intrabond</th>\n",
       "      <th>total sulphide bonds</th>\n",
       "      <th>total glyco positions</th>\n",
       "      <th>first_position_occurance</th>\n",
       "      <th>last_position_occurance</th>\n",
       "      <th>average_Intrabond</th>\n",
       "      <th>avg_Inside_Pairs_Length</th>\n",
       "      <th>glyco_outside_bond</th>\n",
       "      <th>intrabond_glyco_outside_bond</th>\n",
       "      <th>average_Intrabond_outside_glyco_bond</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P04439</td>\n",
       "      <td>HLAA_HUMAN</td>\n",
       "      <td>HLA class I histocompatibility antigen, A alph...</td>\n",
       "      <td>HLA-A HLAA</td>\n",
       "      <td>365</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>40,841</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>[125..188, 227..283]</td>\n",
       "      <td>[110]</td>\n",
       "      <td>...</td>\n",
       "      <td>63|56|</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>125.0</td>\n",
       "      <td>283.0</td>\n",
       "      <td>59.500000</td>\n",
       "      <td>0</td>\n",
       "      <td>[125..188, 227..283]</td>\n",
       "      <td>[63, 56]</td>\n",
       "      <td>59.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P01911</td>\n",
       "      <td>DRB_HUMAN</td>\n",
       "      <td>HLA class II histocompatibility antigen, DRB1 ...</td>\n",
       "      <td>HLA-DRB1</td>\n",
       "      <td>266</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>29,966</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>[44..108, 146..202]</td>\n",
       "      <td>[48]</td>\n",
       "      <td>...</td>\n",
       "      <td>64|56|</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>44.0</td>\n",
       "      <td>202.0</td>\n",
       "      <td>60.000000</td>\n",
       "      <td>64</td>\n",
       "      <td>[nil, 146..202]</td>\n",
       "      <td>[56]</td>\n",
       "      <td>56.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P01889</td>\n",
       "      <td>HLAB_HUMAN</td>\n",
       "      <td>HLA class I histocompatibility antigen, B alph...</td>\n",
       "      <td>HLA-B HLAB</td>\n",
       "      <td>362</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>40,460</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>[125..188, 227..283]</td>\n",
       "      <td>[110]</td>\n",
       "      <td>...</td>\n",
       "      <td>63|56|</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>125.0</td>\n",
       "      <td>283.0</td>\n",
       "      <td>59.500000</td>\n",
       "      <td>0</td>\n",
       "      <td>[125..188, 227..283]</td>\n",
       "      <td>[63, 56]</td>\n",
       "      <td>59.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P08246</td>\n",
       "      <td>ELNE_HUMAN</td>\n",
       "      <td>Neutrophil elastase (EC 3.4.21.37) (Bone marro...</td>\n",
       "      <td>ELANE ELA2</td>\n",
       "      <td>267</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>28,518</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>[55..71, 151..208, 181..187, 198..223]</td>\n",
       "      <td>[88, 124, 173]</td>\n",
       "      <td>...</td>\n",
       "      <td>16|57|6|25|</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>55.0</td>\n",
       "      <td>223.0</td>\n",
       "      <td>26.000000</td>\n",
       "      <td>57</td>\n",
       "      <td>[55..71, nil, 181..187, 198..223]</td>\n",
       "      <td>[16, 6, 25]</td>\n",
       "      <td>15.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P10144</td>\n",
       "      <td>GRAB_HUMAN</td>\n",
       "      <td>Granzyme B (EC 3.4.21.79) (C11) (CTLA-1) (Cath...</td>\n",
       "      <td>GZMB CGL1 CSPB CTLA1 GRB</td>\n",
       "      <td>247</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>27,716</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>[49..65, 142..209, 173..188]</td>\n",
       "      <td>[71, 104]</td>\n",
       "      <td>...</td>\n",
       "      <td>16|67|15|</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>49.0</td>\n",
       "      <td>188.0</td>\n",
       "      <td>32.666667</td>\n",
       "      <td>0</td>\n",
       "      <td>[49..65, 142..209, 173..188]</td>\n",
       "      <td>[16, 67, 15]</td>\n",
       "      <td>32.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2763</th>\n",
       "      <td>A8MPY1</td>\n",
       "      <td>GBRR3_HUMAN</td>\n",
       "      <td>Gamma-aminobutyric acid receptor subunit rho-3...</td>\n",
       "      <td>GABRR3</td>\n",
       "      <td>467</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>54,272</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>[184..198]</td>\n",
       "      <td>[220]</td>\n",
       "      <td>...</td>\n",
       "      <td>14|</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>184.0</td>\n",
       "      <td>198.0</td>\n",
       "      <td>14.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>[184..198]</td>\n",
       "      <td>[14]</td>\n",
       "      <td>14.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2764</th>\n",
       "      <td>P28799</td>\n",
       "      <td>GRN_HUMAN</td>\n",
       "      <td>Progranulin (PGRN) (Acrogranin) (Epithelin pre...</td>\n",
       "      <td>GRN</td>\n",
       "      <td>593</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>63,544</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>[126..139, 133..149, 284..296, 290..306, 297.....</td>\n",
       "      <td>[118, 236, 265, 368, 530]</td>\n",
       "      <td>...</td>\n",
       "      <td>13|16|12|16|17|14|13|13|12|16|13|12|</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "      <td>126.0</td>\n",
       "      <td>416.0</td>\n",
       "      <td>13.916667</td>\n",
       "      <td>12</td>\n",
       "      <td>[126..139, 133..149, 284..296, 290..306, 297.....</td>\n",
       "      <td>[13, 16, 12, 16, 17, 14, 13, 13, 16, 13, 12]</td>\n",
       "      <td>14.090909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2765</th>\n",
       "      <td>O00222</td>\n",
       "      <td>GRM8_HUMAN</td>\n",
       "      <td>Metabotropic glutamate receptor 8 (mGluR8)</td>\n",
       "      <td>GRM8 GPRC1H MGLUR8</td>\n",
       "      <td>908</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>101,741</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>[64..106, 246..534, 369..384, 424..431, 516..5...</td>\n",
       "      <td>[95, 298, 452, 480, 565]</td>\n",
       "      <td>...</td>\n",
       "      <td>42|288|15|7|19|18|12|13|</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>64.0</td>\n",
       "      <td>569.0</td>\n",
       "      <td>51.750000</td>\n",
       "      <td>183</td>\n",
       "      <td>[nil, nil, 369..384, 424..431, 516..535, 520.....</td>\n",
       "      <td>[15, 7, 19, 18, 12]</td>\n",
       "      <td>14.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2766</th>\n",
       "      <td>Q8IXB1</td>\n",
       "      <td>DJC10_HUMAN</td>\n",
       "      <td>DnaJ homolog subfamily C member 10 (EC 1.8.4.-...</td>\n",
       "      <td>DNAJC10 ERDJ5 UNQ495/PRO1012</td>\n",
       "      <td>793</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>91,080</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>[158..161, 480..483, 588..591, 700..703]</td>\n",
       "      <td>[530]</td>\n",
       "      <td>...</td>\n",
       "      <td>3|3|3|3|</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>158.0</td>\n",
       "      <td>703.0</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>[158..161, 480..483, 588..591, 700..703]</td>\n",
       "      <td>[3, 3, 3, 3]</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2767</th>\n",
       "      <td>O95395</td>\n",
       "      <td>GCNT3_HUMAN</td>\n",
       "      <td>Beta-1,3-galactosyl-O-glycosyl-glycoprotein be...</td>\n",
       "      <td>GCNT3</td>\n",
       "      <td>438</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>50,864</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>[70..227, 161..382, 182..209, 391..423]</td>\n",
       "      <td>[289]</td>\n",
       "      <td>...</td>\n",
       "      <td>157|221|27|32|</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>70.0</td>\n",
       "      <td>423.0</td>\n",
       "      <td>109.250000</td>\n",
       "      <td>221</td>\n",
       "      <td>[70..227, nil, 182..209, 391..423]</td>\n",
       "      <td>[157, 27, 32]</td>\n",
       "      <td>72.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2768 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Entry   Entry name                                      Protein names  \\\n",
       "0     P04439   HLAA_HUMAN  HLA class I histocompatibility antigen, A alph...   \n",
       "1     P01911    DRB_HUMAN  HLA class II histocompatibility antigen, DRB1 ...   \n",
       "2     P01889   HLAB_HUMAN  HLA class I histocompatibility antigen, B alph...   \n",
       "3     P08246   ELNE_HUMAN  Neutrophil elastase (EC 3.4.21.37) (Bone marro...   \n",
       "4     P10144   GRAB_HUMAN  Granzyme B (EC 3.4.21.79) (C11) (CTLA-1) (Cath...   \n",
       "...      ...          ...                                                ...   \n",
       "2763  A8MPY1  GBRR3_HUMAN  Gamma-aminobutyric acid receptor subunit rho-3...   \n",
       "2764  P28799    GRN_HUMAN  Progranulin (PGRN) (Acrogranin) (Epithelin pre...   \n",
       "2765  O00222   GRM8_HUMAN         Metabotropic glutamate receptor 8 (mGluR8)   \n",
       "2766  Q8IXB1  DJC10_HUMAN  DnaJ homolog subfamily C member 10 (EC 1.8.4.-...   \n",
       "2767  O95395  GCNT3_HUMAN  Beta-1,3-galactosyl-O-glycosyl-glycoprotein be...   \n",
       "\n",
       "                        Gene names  Length              Organism     Mass  \\\n",
       "0                       HLA-A HLAA     365  Homo sapiens (Human)   40,841   \n",
       "1                         HLA-DRB1     266  Homo sapiens (Human)   29,966   \n",
       "2                       HLA-B HLAB     362  Homo sapiens (Human)   40,460   \n",
       "3                       ELANE ELA2     267  Homo sapiens (Human)   28,518   \n",
       "4         GZMB CGL1 CSPB CTLA1 GRB     247  Homo sapiens (Human)   27,716   \n",
       "...                            ...     ...                   ...      ...   \n",
       "2763                        GABRR3     467  Homo sapiens (Human)   54,272   \n",
       "2764                           GRN     593  Homo sapiens (Human)   63,544   \n",
       "2765            GRM8 GPRC1H MGLUR8     908  Homo sapiens (Human)  101,741   \n",
       "2766  DNAJC10 ERDJ5 UNQ495/PRO1012     793  Homo sapiens (Human)   91,080   \n",
       "2767                         GCNT3     438  Homo sapiens (Human)   50,864   \n",
       "\n",
       "        Status                                     Disulfide bond  \\\n",
       "0     reviewed                               [125..188, 227..283]   \n",
       "1     reviewed                                [44..108, 146..202]   \n",
       "2     reviewed                               [125..188, 227..283]   \n",
       "3     reviewed             [55..71, 151..208, 181..187, 198..223]   \n",
       "4     reviewed                       [49..65, 142..209, 173..188]   \n",
       "...        ...                                                ...   \n",
       "2763  reviewed                                         [184..198]   \n",
       "2764  reviewed  [126..139, 133..149, 284..296, 290..306, 297.....   \n",
       "2765  reviewed  [64..106, 246..534, 369..384, 424..431, 516..5...   \n",
       "2766  reviewed           [158..161, 480..483, 588..591, 700..703]   \n",
       "2767  reviewed            [70..227, 161..382, 182..209, 391..423]   \n",
       "\n",
       "                  Glycosylation  ...                             intrabond  \\\n",
       "0                         [110]  ...                                63|56|   \n",
       "1                          [48]  ...                                64|56|   \n",
       "2                         [110]  ...                                63|56|   \n",
       "3                [88, 124, 173]  ...                           16|57|6|25|   \n",
       "4                     [71, 104]  ...                             16|67|15|   \n",
       "...                         ...  ...                                   ...   \n",
       "2763                      [220]  ...                                   14|   \n",
       "2764  [118, 236, 265, 368, 530]  ...  13|16|12|16|17|14|13|13|12|16|13|12|   \n",
       "2765   [95, 298, 452, 480, 565]  ...              42|288|15|7|19|18|12|13|   \n",
       "2766                      [530]  ...                              3|3|3|3|   \n",
       "2767                      [289]  ...                        157|221|27|32|   \n",
       "\n",
       "     total sulphide bonds total glyco positions  first_position_occurance  \\\n",
       "0                       2                     1                     125.0   \n",
       "1                       2                     1                      44.0   \n",
       "2                       2                     1                     125.0   \n",
       "3                       4                     3                      55.0   \n",
       "4                       3                     2                      49.0   \n",
       "...                   ...                   ...                       ...   \n",
       "2763                    1                     1                     184.0   \n",
       "2764                   12                     5                     126.0   \n",
       "2765                    8                     5                      64.0   \n",
       "2766                    4                     1                     158.0   \n",
       "2767                    4                     1                      70.0   \n",
       "\n",
       "     last_position_occurance average_Intrabond  avg_Inside_Pairs_Length  \\\n",
       "0                      283.0         59.500000                        0   \n",
       "1                      202.0         60.000000                       64   \n",
       "2                      283.0         59.500000                        0   \n",
       "3                      223.0         26.000000                       57   \n",
       "4                      188.0         32.666667                        0   \n",
       "...                      ...               ...                      ...   \n",
       "2763                   198.0         14.000000                        0   \n",
       "2764                   416.0         13.916667                       12   \n",
       "2765                   569.0         51.750000                      183   \n",
       "2766                   703.0          3.000000                        0   \n",
       "2767                   423.0        109.250000                      221   \n",
       "\n",
       "                                     glyco_outside_bond  \\\n",
       "0                                  [125..188, 227..283]   \n",
       "1                                       [nil, 146..202]   \n",
       "2                                  [125..188, 227..283]   \n",
       "3                     [55..71, nil, 181..187, 198..223]   \n",
       "4                          [49..65, 142..209, 173..188]   \n",
       "...                                                 ...   \n",
       "2763                                         [184..198]   \n",
       "2764  [126..139, 133..149, 284..296, 290..306, 297.....   \n",
       "2765  [nil, nil, 369..384, 424..431, 516..535, 520.....   \n",
       "2766           [158..161, 480..483, 588..591, 700..703]   \n",
       "2767                 [70..227, nil, 182..209, 391..423]   \n",
       "\n",
       "                      intrabond_glyco_outside_bond  \\\n",
       "0                                         [63, 56]   \n",
       "1                                             [56]   \n",
       "2                                         [63, 56]   \n",
       "3                                      [16, 6, 25]   \n",
       "4                                     [16, 67, 15]   \n",
       "...                                            ...   \n",
       "2763                                          [14]   \n",
       "2764  [13, 16, 12, 16, 17, 14, 13, 13, 16, 13, 12]   \n",
       "2765                           [15, 7, 19, 18, 12]   \n",
       "2766                                  [3, 3, 3, 3]   \n",
       "2767                                 [157, 27, 32]   \n",
       "\n",
       "      average_Intrabond_outside_glyco_bond  \n",
       "0                                59.500000  \n",
       "1                                56.000000  \n",
       "2                                59.500000  \n",
       "3                                15.666667  \n",
       "4                                32.666667  \n",
       "...                                    ...  \n",
       "2763                             14.000000  \n",
       "2764                             14.090909  \n",
       "2765                             14.200000  \n",
       "2766                              3.000000  \n",
       "2767                             72.000000  \n",
       "\n",
       "[2768 rows x 25 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
