{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re \n",
    "import matplotlib.pyplot as plt \n",
    "import sys\n",
    "import time\n",
    "\n",
    "\n",
    "\n",
    "#Input file must be given to run the code.\n",
    "#For Example Run(in terminal) 'python3 Bioinformatics input.xlsx'\n",
    "\n",
    "#Filename =sys.argv[1] # Read input file \n",
    "\n",
    "df = pd.read_excel('updatedBigDataInput.xlsx')# Read EXCEL Files and storing in database 'df'\n",
    "\n",
    "data= pd.DataFrame()# Creating New Data Frame \n",
    "\n",
    "glyco_sulfide_frame= pd.DataFrame()   #Data Frame which contains Glycosilation and sulfide.\n",
    "\n",
    "# Storing Usefull information From the database 'df' \n",
    "data['Entry']=df['Entry'] #Storing Entry\n",
    "data['Entry name']=df['Entry name']  #Storing Entry Name\n",
    "data['Protein names']=df['Protein names']  #Storing Protein Names\n",
    "data['Gene names']=df['Gene names'] # Storing Gene Names\n",
    "data['Length']=df['Length'] # Storing Length\n",
    "data['Organism']=df['Organism'] # Storing Type of Organism\n",
    "data['Mass']=df['Mass'] #Storing Mass\n",
    "data['Status']=df['Status']\n",
    "\n",
    "#Now removing the extra data we dont need and containg only organism 'Human'\n",
    "human_data=data[data['Organism'].str.contains(\"Homo sapiens\")] #Selecting only organism which contains 'Homo sapiens'\n",
    "data= human_data\n",
    "\n",
    "data['Disulfide bond']=df['Disulfide bond'] # Storing Disulfide bond \n",
    "data['Glycosylation']= df['Glycosylation']# Storing Glycosylation Position\n",
    "#data=data.dropna().reset_index(drop=True) # Drop all rows which contains Not a number and reset the index \n",
    "\n",
    "# Working on getting the relative positions of the Disulfide bond\n",
    "temp_disulfide= data['Disulfide bond'] # Storing data temporary as 'disulfide_column'\n",
    "disulfide_column= data['Disulfide bond']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "expected string or bytes-like object",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-6-b4bf25db6ce5>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      2\u001b[0m     \u001b[1;32mreturn\u001b[0m \u001b[0mre\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfindall\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'\\d+\\..\\d+'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mnewdata\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 4\u001b[1;33m \u001b[0mbond\u001b[0m\u001b[1;33m=\u001b[0m \u001b[0mtemp_disulfide\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mget_sulfide_value\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;31m# Function call which gives all the positons\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      5\u001b[0m \u001b[0mglyco_sulfide_frame\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'bond'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mbond\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      6\u001b[0m \u001b[0mdata\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'Disulfide bond'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mdisulfide_column\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mget_sulfide_value\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;31m# Storing extracted disulfide value in data Frame.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\pandas\\core\\series.py\u001b[0m in \u001b[0;36mapply\u001b[1;34m(self, func, convert_dtype, args, **kwds)\u001b[0m\n\u001b[0;32m   4040\u001b[0m             \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   4041\u001b[0m                 \u001b[0mvalues\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mobject\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 4042\u001b[1;33m                 \u001b[0mmapped\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mlib\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmap_infer\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mf\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mconvert\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mconvert_dtype\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   4043\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   4044\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmapped\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mand\u001b[0m \u001b[0misinstance\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmapped\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mSeries\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mpandas\\_libs\\lib.pyx\u001b[0m in \u001b[0;36mpandas._libs.lib.map_infer\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32m<ipython-input-6-b4bf25db6ce5>\u001b[0m in \u001b[0;36mget_sulfide_value\u001b[1;34m(newdata)\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[1;32mdef\u001b[0m \u001b[0mget_sulfide_value\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mnewdata\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 2\u001b[1;33m     \u001b[1;32mreturn\u001b[0m \u001b[0mre\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfindall\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'\\d+\\..\\d+'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mnewdata\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      3\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[0mbond\u001b[0m\u001b[1;33m=\u001b[0m \u001b[0mtemp_disulfide\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mget_sulfide_value\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;31m# Function call which gives all the positons\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m \u001b[0mglyco_sulfide_frame\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'bond'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mbond\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\re.py\u001b[0m in \u001b[0;36mfindall\u001b[1;34m(pattern, string, flags)\u001b[0m\n\u001b[0;32m    221\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    222\u001b[0m     Empty matches are included in the result.\"\"\"\n\u001b[1;32m--> 223\u001b[1;33m     \u001b[1;32mreturn\u001b[0m \u001b[0m_compile\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mpattern\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mflags\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfindall\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mstring\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    224\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    225\u001b[0m \u001b[1;32mdef\u001b[0m \u001b[0mfinditer\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mpattern\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mstring\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mflags\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mTypeError\u001b[0m: expected string or bytes-like object"
     ]
    }
   ],
   "source": [
    "\n",
    "def get_sulfide_value(newdata):\n",
    "    return re.findall('\\d+\\..\\d+',newdata)\n",
    "\n",
    "bond= temp_disulfide.apply(get_sulfide_value) # Function call which gives all the positons \n",
    "glyco_sulfide_frame['bond']=bond\n",
    "data['Disulfide bond']=disulfide_column.apply(get_sulfide_value)# Storing extracted disulfide value in data Frame.\n",
    "\n",
    "\n",
    "def get_int_value(data):        #function defined for global purpose use to get integer values for any column.\n",
    "    return re.findall('\\d+',data)\n",
    "def get_float_value(data):\n",
    "    return re.findall('\\d+\\.\\d+',data)\n",
    "\n",
    "\n",
    "def get_Nlinked(data):\n",
    "    return re.findall('\\d+\\;\\s*\\/note=\\\"N-linked',data)\n",
    "    \n",
    "Glyco_data= data['Glycosylation'] # Storing data temporary as 'Glyco_data'\n",
    "data['Glycosylation']= Glyco_data.apply(get_Nlinked) # the return data will be like 'CARBOHYD 110'\n",
    "glyco_sulfide_frame['glyco']=data['Glycosylation']\n",
    "data=data.dropna().reset_index(drop=True)\n",
    "glyco_sulfide_frame=glyco_sulfide_frame.dropna().reset_index(drop=True)\n",
    "\n",
    "#Now Removing the extra word 'CARBOHYD' and getting all positions of Glycosylation.\n",
    "temp_data= data['Glycosylation'].astype(str)\n",
    "data['Glycosylation']= temp_data.apply(get_int_value) \n",
    "\n",
    "\n",
    "\n",
    "# Creating a temporary dataframe for total length in columns and length in rows(like number of pairs in each index).\n",
    "a= pd.DataFrame()\n",
    "b= pd.DataFrame()\n",
    "a =pd.Series(data['Glycosylation'])\n",
    "b =pd.Series(data['Disulfide bond'])\n",
    "lenA=a.str.len() #length of Glycosylation\n",
    "lenB=b.str.len() # length of Disulfide\n",
    "\n",
    "\n",
    "rel_pos = pd.Series(a.size,dtype=np.str)\n",
    "for i in range(a.size):\n",
    "    rel_pos[i]=''\n",
    "    \n",
    "InsidePairs=pd.Series(b.size,dtype=np.str)\n",
    "InsidePairsLength=np.zeros(b.size,dtype=np.int)\n",
    "for i in range(a.size):\n",
    "    InsidePairs[i]=''\n",
    "\n",
    "# Now using loop and properties of Python library 're' to get values and finding positions between disulphide bond and glycosation.\n",
    "i=0\n",
    "while i<b.size:\n",
    "    for j in range(lenA[i]):\n",
    "        Position=int(a[i][j])\n",
    "        rel_pos[i]=rel_pos[i]+str(a[i][j])+'{|'\n",
    "        for k in range(lenB[i]):\n",
    "            split = re.findall('\\d+',b[i][k]) \n",
    "            firstPair= int(split[0])\n",
    "            secondPair=int(split[1])\n",
    "            if Position< firstPair:\n",
    "                rel_pos[i] = rel_pos[i]+'o'+str(Position-firstPair)\n",
    "                rel_pos[i] = rel_pos[i]+'o'+str(Position-secondPair)\n",
    "            elif Position>= firstPair and Position<=secondPair:\n",
    "                rel_pos[i]=  rel_pos[i]+'i' + str(Position-firstPair)\n",
    "                rel_pos[i] = rel_pos[i]+'i'+ str(Position-secondPair)\n",
    "                InsidePairs[i]= InsidePairs[i] + str('('+ str(firstPair)+','+str(secondPair)+')')\n",
    "                InsidePairsLength[i]=InsidePairsLength[i]+ (secondPair-firstPair)\n",
    "            else:\n",
    "                rel_pos[i] = rel_pos[i]+'o'+str(Position-firstPair)\n",
    "                rel_pos[i] = rel_pos[i]+'o'+str(Position-secondPair)\n",
    "            rel_pos[i]=rel_pos[i]+'|'    \n",
    "        rel_pos[i]=rel_pos[i]+'}'\n",
    "    i=i+1\n",
    "data['rel_pos']= rel_pos\n",
    "data['In_Gly_Dis_Pair']=InsidePairs\n",
    "\n",
    "#Caluclating Score of each pairs of disulfide bonds\n",
    "# O score if other pair outside.\n",
    "# 0.5 score if other pair half inside and half outside.\n",
    "# 1.0 score if other pair full inside.\n",
    "array = pd.Series(a.size,dtype=np.str)\n",
    "for x in range(a.size):\n",
    "    array[x]=''\n",
    "\n",
    "for i in range(0,len(data)):\n",
    "    temp_array = np.zeros(lenB[i])\n",
    "    for j in range(0,lenB[i]):\n",
    "        for k in range(j+1,lenB[i]):\n",
    "            if k!=lenB[i]:\n",
    "                split1 = re.findall('\\d+',data['Disulfide bond'][i][j]) \n",
    "                p1e1=int(split1[0])\n",
    "                p1e2=int(split1[1])\n",
    "                split2 = re.findall('\\d+',data['Disulfide bond'][i][k]) \n",
    "                p2e1=int(split2[0])\n",
    "                p2e2=int(split2[1])\n",
    "            \n",
    "                if(p2e1>p1e2):\n",
    "                    temp_array[j]=temp_array[j]+0\n",
    "                    temp_array[k]=temp_array[k]+0\n",
    "                elif (p2e1<p1e2) and (p2e2>p1e2):\n",
    "                    temp_array[j]=temp_array[j]+0.5\n",
    "                    temp_array[k]=temp_array[k]+0.5    \n",
    "                elif (p2e1<p1e2) and (p2e2<p1e2):\n",
    "                    temp_array[j]=temp_array[j]+1.0\n",
    "                    temp_array[k]=temp_array[k]+0.0\n",
    "        array[i]=array[i]+str(temp_array[j])+','\n",
    "data['Disulfide_score']= array.apply(get_float_value)\n",
    "\n",
    "#Claculating disulfide pair score average \n",
    "disulfie_pair_score_sum = np.zeros(len(data),dtype=np.float)\n",
    "for i in range(len(data)):\n",
    "    for j in range(len(data['Disulfide_score'][i])):\n",
    "        disulfie_pair_score_sum[i]= disulfie_pair_score_sum[i]+ float(data['Disulfide_score'][i][j])\n",
    "data['disulfie_pair_score_sum']=disulfie_pair_score_sum\n",
    "\n",
    "\n",
    "## Calculating Interbond Distance\n",
    "interbond_distance= pd.Series(a.size,dtype= np.str)\n",
    "\n",
    "for i in range(a.size):\n",
    "    interbond_distance[i]=''\n",
    "\n",
    "for i in range(a.size):\n",
    "    \n",
    "    if lenB[i]==0:\n",
    "        interbond_distance[i]='No pair'\n",
    "        \n",
    "    if lenB[i]==1:\n",
    "        interbond_distance[i]='Only One pair'\n",
    "    \n",
    "    for r in range(lenB[i]-1):\n",
    "        split1= re.findall('\\d+',data['Disulfide bond'][i][r])\n",
    "        split2= re.findall('\\d+',data['Disulfide bond'][i][r+1])\n",
    "        interbond_distance[i]= interbond_distance[i]+ str(int(split2[0])-int(split1[1]))+ '|'\n",
    "        \n",
    "data['interbond_distance']=interbond_distance\n",
    "\n",
    "#Intrabond distance formation\n",
    "intrabond=pd.Series(b.size,dtype=np.str)\n",
    "for i in range(a.size):\n",
    "    intrabond[i]=''\n",
    "j=0;\n",
    "while j<b.size:\n",
    "    for k in range(lenB[j]):\n",
    "        split = re.findall('\\d+',b[j][k]) \n",
    "        firstPair= int(split[0])\n",
    "        secondPair=int(split[1])\n",
    "        intrabond[j]=intrabond[j]+str(secondPair-firstPair)+'|'    \n",
    "    j=j+1\n",
    "data['intrabond']=intrabond\n",
    "\n",
    "\n",
    "#number of pairs of disuphide bonds\n",
    "data['total sulphide bonds']=lenB\n",
    "\n",
    "#number of glcosylation positions\n",
    "data['total glyco positions']=lenA\n",
    "\n",
    "\n",
    "#1st disulphide pair to N-terminus distance\n",
    "first_pos = pd.Series(a.size,dtype=np.int)\n",
    "for x in range(a.size):\n",
    "    if(data['Disulfide bond'][x]!=[]):\n",
    "        split = re.findall('\\d+',data['Disulfide bond'][x][0]) \n",
    "        firstPair= int(split[0])\n",
    "        first_pos[x]= firstPair\n",
    "data['first_position_occurance']=first_pos\n",
    "\n",
    "\n",
    "#last disulphide pair to C-terminus distance\n",
    "last_pos= pd.Series(a.size,dtype=np.int)\n",
    "for x in range(a.size):\n",
    "    if(data['Disulfide bond'][x]!=[]):\n",
    "        split= re.findall('\\d+',data['Disulfide bond'][x][(lenB[x]-1)])\n",
    "        lastPair= int(split[1])\n",
    "        last_pos[x] = lastPair \n",
    "data['last_position_occurance']=last_pos\n",
    "\n",
    "#Calculate average distance of intrabond distance\n",
    "average= data['intrabond'].apply(get_int_value)\n",
    "lenIntra=average.str.len()\n",
    "\n",
    "averageintrabond= np.zeros(a.size)\n",
    "for i in range(a.size):\n",
    "    totalsum=0\n",
    "    for j in range(lenIntra[i]):\n",
    "        totalsum= totalsum+int(average[i][j])\n",
    "    averageintrabond[i]=totalsum/lenIntra[i]\n",
    "data['average_Intrabond']=averageintrabond\n",
    "\n",
    "\n",
    "#calculating average distance of those pairs of disulphide bonds that have Glycosylation Inside.\n",
    "temp_length=pd.Series(b.size,dtype=np.str)\n",
    "#for i in range(a.size):\n",
    "#    temp_length[i]=''\n",
    "inside_length = np.zeros(a.size,dtype=np.int)\n",
    "for r in range(a.size):\n",
    "    inside_length[r]=len(re.findall('\\d+',data['In_Gly_Dis_Pair'][r]))/2\n",
    "    temp_length[r]=re.findall('\\d+',data['In_Gly_Dis_Pair'][r])\n",
    "    if inside_length[r]!=0:\n",
    "        InsidePairsLength[r]=InsidePairsLength[r]/inside_length[r]\n",
    "data['avg_Inside_Pairs_Length']=InsidePairsLength\n",
    "\n",
    "\n",
    "#If the glyco is inside then putting the sulphide position 'nil' at that point.\n",
    "i=0\n",
    "while i<b.size:\n",
    "    for j in range(lenA[i]):\n",
    "        Position=int(a[i][j])\n",
    "        for k in range(lenB[i]):\n",
    "            if glyco_sulfide_frame['bond'][i][k]!='nil':\n",
    "                split = re.findall('\\d+',b[i][k]) \n",
    "                firstPair= int(split[0])\n",
    "                secondPair=int(split[1])\n",
    "                if Position>= firstPair and Position<=secondPair:\n",
    "                    glyco_sulfide_frame['bond'][i][k]='nil'\n",
    "    i=i+1\n",
    "\n",
    "\n",
    "#Calculating length between sulfide bonds.\n",
    "temp=pd.Series(b.size,dtype=np.str)\n",
    "for i in range(a.size):\n",
    "    temp[i]=''\n",
    "\n",
    "for i in range(len(glyco_sulfide_frame['bond'])):\n",
    "    for j in range(lenB[i]):\n",
    "        if glyco_sulfide_frame['bond'][i][j]!='nil':\n",
    "            split= re.findall('\\d+',glyco_sulfide_frame['bond'][i][j])\n",
    "            temp[i]=str(temp[i] + str(int(split[1])-int(split[0]))+'|')\n",
    "data['glyco_outside_bond']=glyco_sulfide_frame['bond']\n",
    "\n",
    "#get the integer values.\n",
    "intrabond_outside= temp.apply(get_int_value)\n",
    "data['intrabond_glyco_outside_bond']=intrabond_outside\n",
    "\n",
    "\n",
    "# In[42]:\n",
    "\n",
    "\n",
    "#calculating avegrage lenth of sulphide bonds which do not have glyco inside.\n",
    "\n",
    "new =pd.Series(intrabond_outside)\n",
    "newlen=new.str.len()\n",
    "average= np.zeros(a.size)\n",
    "for i in range(a.size):\n",
    "    totalsum=0\n",
    "    for j in range(newlen[i]):\n",
    "        totalsum= totalsum+int(intrabond_outside[i][j])\n",
    "    if newlen[i]!=0:\n",
    "        average[i]=totalsum/newlen[i]\n",
    "data['average_Intrabond_outside_glyco_bond']=average\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "expected string or bytes-like object",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-3-00bdc9fda099>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      9\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     10\u001b[0m \u001b[0mdata1\u001b[0m\u001b[1;33m=\u001b[0m \u001b[0mdata\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'Glycosylation'\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;31m# Storing data temporary as 'Glyco_data'\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 11\u001b[1;33m \u001b[0mdata\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'Glycosylation_GlcNac'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m=\u001b[0m \u001b[0mdata1\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mget_Nlinked_GlcNAc\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;31m# the return data with 'N-linked (GlcNAc...)'\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     12\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     13\u001b[0m \u001b[0mdata2\u001b[0m\u001b[1;33m=\u001b[0m \u001b[0mdata\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'Glycosylation'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\pandas\\core\\series.py\u001b[0m in \u001b[0;36mapply\u001b[1;34m(self, func, convert_dtype, args, **kwds)\u001b[0m\n\u001b[0;32m   4040\u001b[0m             \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   4041\u001b[0m                 \u001b[0mvalues\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mobject\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 4042\u001b[1;33m                 \u001b[0mmapped\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mlib\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmap_infer\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mf\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mconvert\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mconvert_dtype\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   4043\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   4044\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmapped\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mand\u001b[0m \u001b[0misinstance\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmapped\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mSeries\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mpandas\\_libs\\lib.pyx\u001b[0m in \u001b[0;36mpandas._libs.lib.map_infer\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32m<ipython-input-3-00bdc9fda099>\u001b[0m in \u001b[0;36mget_Nlinked_GlcNAc\u001b[1;34m(data)\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[1;32mdef\u001b[0m \u001b[0mget_Nlinked_GlcNAc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 2\u001b[1;33m     \u001b[1;32mreturn\u001b[0m \u001b[0mre\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfindall\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'\\d+;  /note=\"N-linked \\(GlcNAc...\\) asparagine\"'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mdata\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      3\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[1;32mdef\u001b[0m \u001b[0mget_Nlinked_Glc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m     \u001b[1;32mreturn\u001b[0m \u001b[0mre\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfindall\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'\\d+;  /note=\"N-linked \\(Glc\\)'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mdata\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\re.py\u001b[0m in \u001b[0;36mfindall\u001b[1;34m(pattern, string, flags)\u001b[0m\n\u001b[0;32m    221\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    222\u001b[0m     Empty matches are included in the result.\"\"\"\n\u001b[1;32m--> 223\u001b[1;33m     \u001b[1;32mreturn\u001b[0m \u001b[0m_compile\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mpattern\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mflags\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfindall\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mstring\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    224\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    225\u001b[0m \u001b[1;32mdef\u001b[0m \u001b[0mfinditer\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mpattern\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mstring\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mflags\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mTypeError\u001b[0m: expected string or bytes-like object"
     ]
    }
   ],
   "source": [
    "def get_Nlinked_GlcNAc(data):\n",
    "    return re.findall('\\d+;  /note=\"N-linked \\(GlcNAc...\\) asparagine\"',data)\n",
    "\n",
    "def get_Nlinked_Glc(data):\n",
    "    return re.findall('\\d+;  /note=\"N-linked \\(Glc\\)',data)\n",
    "\n",
    "def get_OLinked(data):\n",
    "    return re.findall('\\d+\\;  /note=\\\"O-linked',data)\n",
    "    \n",
    "data1= data['Glycosylation'] # Storing data temporary as 'Glyco_data'\n",
    "data['Glycosylation_GlcNac']= data1.apply(get_Nlinked_GlcNAc) # the return data with 'N-linked (GlcNAc...)'\n",
    "\n",
    "data2= data['Glycosylation']\n",
    "data['Glycosylation_Nlinked_Glc']=data2.apply(get_Nlinked_Glc)\n",
    "\n",
    "data3= data['Glycosylation']\n",
    "data['Glycosylation_Olinked']=data3.apply(get_OLinked)\n",
    "\n",
    "#Now Removing the extra word 'CARBOHYD' and getting all positions of Glycosylation.\n",
    "temp_data= data['Glycosylation_GlcNac'].astype(str)\n",
    "data['Glycosylation_GlcNac']= temp_data.apply(get_int_value) \n",
    "\n",
    "temp_data= data['Glycosylation_Nlinked_Glc'].astype(str)\n",
    "data['Glycosylation_Nlinked_Glc']= temp_data.apply(get_int_value) \n",
    "\n",
    "temp_data= data['Glycosylation_Olinked'].astype(str)\n",
    "data['Glycosylation_Olinked']= temp_data.apply(get_int_value) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data.to_excel('finaloutput.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry name</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene names</th>\n",
       "      <th>Length</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Mass</th>\n",
       "      <th>Status</th>\n",
       "      <th>Disulfide bond</th>\n",
       "      <th>Glycosylation</th>\n",
       "      <th>Filtered</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>P04439</td>\n",
       "      <td>HLAA_HUMAN</td>\n",
       "      <td>HLA class I histocompatibility antigen, A alph...</td>\n",
       "      <td>HLA-A HLAA</td>\n",
       "      <td>365</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>40,841</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>DISULFID 125..188;  /evidence=\"ECO:0000255|PRO...</td>\n",
       "      <td>CARBOHYD 110;  /note=\"N-linked (GlcNAc...) asp...</td>\n",
       "      <td>CARBOHYD 110;  /note=\"N-linked (GlcNAc...) asp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>P31689</td>\n",
       "      <td>DNJA1_HUMAN</td>\n",
       "      <td>DnaJ homolog subfamily A member 1 (DnaJ protei...</td>\n",
       "      <td>DNAJA1 DNAJ2 HDJ2 HSJ2 HSPF4</td>\n",
       "      <td>397</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>44,868</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>P01911</td>\n",
       "      <td>DRB_HUMAN</td>\n",
       "      <td>HLA class II histocompatibility antigen, DRB1 ...</td>\n",
       "      <td>HLA-DRB1</td>\n",
       "      <td>266</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>29,966</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>DISULFID 44..108;  /evidence=\"ECO:0000269|PubM...</td>\n",
       "      <td>CARBOHYD 48;  /note=\"N-linked (GlcNAc...) aspa...</td>\n",
       "      <td>CARBOHYD 48;  /note=\"N-linked (GlcNAc...) aspa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>P01889</td>\n",
       "      <td>HLAB_HUMAN</td>\n",
       "      <td>HLA class I histocompatibility antigen, B alph...</td>\n",
       "      <td>HLA-B HLAB</td>\n",
       "      <td>362</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>40,460</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>DISULFID 125..188;  /evidence=\"ECO:0000255|PRO...</td>\n",
       "      <td>CARBOHYD 110;  /note=\"N-linked (GlcNAc...) asp...</td>\n",
       "      <td>CARBOHYD 110;  /note=\"N-linked (GlcNAc...) asp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>P08246</td>\n",
       "      <td>ELNE_HUMAN</td>\n",
       "      <td>Neutrophil elastase (EC 3.4.21.37) (Bone marro...</td>\n",
       "      <td>ELANE ELA2</td>\n",
       "      <td>267</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>28,518</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>DISULFID 55..71;  /evidence=\"ECO:0000255|PROSI...</td>\n",
       "      <td>CARBOHYD 88;  /note=\"N-linked (GlcNAc...) aspa...</td>\n",
       "      <td>CARBOHYD 88;  /note=\"N-linked (GlcNAc...) aspa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>49795</td>\n",
       "      <td>Q9NZH0</td>\n",
       "      <td>GPC5B_HUMAN</td>\n",
       "      <td>G-protein coupled receptor family C group 5 me...</td>\n",
       "      <td>GPRC5B RAIG2</td>\n",
       "      <td>403</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>44,795</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CARBOHYD 30;  /note=\"N-linked (GlcNAc...) aspa...</td>\n",
       "      <td>CARBOHYD 30;  /note=\"N-linked (GlcNAc...) aspa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>49796</td>\n",
       "      <td>Q9NXW2</td>\n",
       "      <td>DJB12_HUMAN</td>\n",
       "      <td>DnaJ homolog subfamily B member 12</td>\n",
       "      <td>DNAJB12</td>\n",
       "      <td>375</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>41,860</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>49798</td>\n",
       "      <td>O95395</td>\n",
       "      <td>GCNT3_HUMAN</td>\n",
       "      <td>Beta-1,3-galactosyl-O-glycosyl-glycoprotein be...</td>\n",
       "      <td>GCNT3</td>\n",
       "      <td>438</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>50,864</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>DISULFID 70..227;  /evidence=\"ECO:0000250\"; DI...</td>\n",
       "      <td>CARBOHYD 289;  /note=\"N-linked (GlcNAc...) asp...</td>\n",
       "      <td>CARBOHYD 289;  /note=\"N-linked (GlcNAc...) asp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>49803</td>\n",
       "      <td>Q9Y238</td>\n",
       "      <td>DLEC1_HUMAN</td>\n",
       "      <td>Deleted in lung and esophageal cancer protein ...</td>\n",
       "      <td>DLEC1 DLC1</td>\n",
       "      <td>1755</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>195,684</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>49805</td>\n",
       "      <td>Q8TDM6</td>\n",
       "      <td>DLG5_HUMAN</td>\n",
       "      <td>Disks large homolog 5 (Discs large protein P-d...</td>\n",
       "      <td>DLG5 KIAA0583 PDLG</td>\n",
       "      <td>1919</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>213,868</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20369 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Entry   Entry name                                      Protein names  \\\n",
       "0      P04439   HLAA_HUMAN  HLA class I histocompatibility antigen, A alph...   \n",
       "1      P31689  DNJA1_HUMAN  DnaJ homolog subfamily A member 1 (DnaJ protei...   \n",
       "2      P01911    DRB_HUMAN  HLA class II histocompatibility antigen, DRB1 ...   \n",
       "3      P01889   HLAB_HUMAN  HLA class I histocompatibility antigen, B alph...   \n",
       "4      P08246   ELNE_HUMAN  Neutrophil elastase (EC 3.4.21.37) (Bone marro...   \n",
       "...       ...          ...                                                ...   \n",
       "49795  Q9NZH0  GPC5B_HUMAN  G-protein coupled receptor family C group 5 me...   \n",
       "49796  Q9NXW2  DJB12_HUMAN                 DnaJ homolog subfamily B member 12   \n",
       "49798  O95395  GCNT3_HUMAN  Beta-1,3-galactosyl-O-glycosyl-glycoprotein be...   \n",
       "49803  Q9Y238  DLEC1_HUMAN  Deleted in lung and esophageal cancer protein ...   \n",
       "49805  Q8TDM6   DLG5_HUMAN  Disks large homolog 5 (Discs large protein P-d...   \n",
       "\n",
       "                         Gene names  Length              Organism     Mass  \\\n",
       "0                        HLA-A HLAA     365  Homo sapiens (Human)   40,841   \n",
       "1      DNAJA1 DNAJ2 HDJ2 HSJ2 HSPF4     397  Homo sapiens (Human)   44,868   \n",
       "2                          HLA-DRB1     266  Homo sapiens (Human)   29,966   \n",
       "3                        HLA-B HLAB     362  Homo sapiens (Human)   40,460   \n",
       "4                        ELANE ELA2     267  Homo sapiens (Human)   28,518   \n",
       "...                             ...     ...                   ...      ...   \n",
       "49795                  GPRC5B RAIG2     403  Homo sapiens (Human)   44,795   \n",
       "49796                       DNAJB12     375  Homo sapiens (Human)   41,860   \n",
       "49798                         GCNT3     438  Homo sapiens (Human)   50,864   \n",
       "49803                    DLEC1 DLC1    1755  Homo sapiens (Human)  195,684   \n",
       "49805            DLG5 KIAA0583 PDLG    1919  Homo sapiens (Human)  213,868   \n",
       "\n",
       "         Status                                     Disulfide bond  \\\n",
       "0      reviewed  DISULFID 125..188;  /evidence=\"ECO:0000255|PRO...   \n",
       "1      reviewed                                                NaN   \n",
       "2      reviewed  DISULFID 44..108;  /evidence=\"ECO:0000269|PubM...   \n",
       "3      reviewed  DISULFID 125..188;  /evidence=\"ECO:0000255|PRO...   \n",
       "4      reviewed  DISULFID 55..71;  /evidence=\"ECO:0000255|PROSI...   \n",
       "...         ...                                                ...   \n",
       "49795  reviewed                                                NaN   \n",
       "49796  reviewed                                                NaN   \n",
       "49798  reviewed  DISULFID 70..227;  /evidence=\"ECO:0000250\"; DI...   \n",
       "49803  reviewed                                                NaN   \n",
       "49805  reviewed                                                NaN   \n",
       "\n",
       "                                           Glycosylation  \\\n",
       "0      CARBOHYD 110;  /note=\"N-linked (GlcNAc...) asp...   \n",
       "1                                                    NaN   \n",
       "2      CARBOHYD 48;  /note=\"N-linked (GlcNAc...) aspa...   \n",
       "3      CARBOHYD 110;  /note=\"N-linked (GlcNAc...) asp...   \n",
       "4      CARBOHYD 88;  /note=\"N-linked (GlcNAc...) aspa...   \n",
       "...                                                  ...   \n",
       "49795  CARBOHYD 30;  /note=\"N-linked (GlcNAc...) aspa...   \n",
       "49796                                                NaN   \n",
       "49798  CARBOHYD 289;  /note=\"N-linked (GlcNAc...) asp...   \n",
       "49803                                                NaN   \n",
       "49805                                                NaN   \n",
       "\n",
       "                                                Filtered  \n",
       "0      CARBOHYD 110;  /note=\"N-linked (GlcNAc...) asp...  \n",
       "1                                                    NaN  \n",
       "2      CARBOHYD 48;  /note=\"N-linked (GlcNAc...) aspa...  \n",
       "3      CARBOHYD 110;  /note=\"N-linked (GlcNAc...) asp...  \n",
       "4      CARBOHYD 88;  /note=\"N-linked (GlcNAc...) aspa...  \n",
       "...                                                  ...  \n",
       "49795  CARBOHYD 30;  /note=\"N-linked (GlcNAc...) aspa...  \n",
       "49796                                                NaN  \n",
       "49798  CARBOHYD 289;  /note=\"N-linked (GlcNAc...) asp...  \n",
       "49803                                                NaN  \n",
       "49805                                                NaN  \n",
       "\n",
       "[20369 rows x 11 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry name</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene names</th>\n",
       "      <th>Length</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Mass</th>\n",
       "      <th>Status</th>\n",
       "      <th>Disulfide bond</th>\n",
       "      <th>Glycosylation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>P04439</td>\n",
       "      <td>HLAA_HUMAN</td>\n",
       "      <td>HLA class I histocompatibility antigen, A alph...</td>\n",
       "      <td>HLA-A HLAA</td>\n",
       "      <td>365</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>40,841</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>DISULFID 125..188;  /evidence=\"ECO:0000255|PRO...</td>\n",
       "      <td>CARBOHYD 110;  /note=\"N-linked (GlcNAc...) asp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>P31689</td>\n",
       "      <td>DNJA1_HUMAN</td>\n",
       "      <td>DnaJ homolog subfamily A member 1 (DnaJ protei...</td>\n",
       "      <td>DNAJA1 DNAJ2 HDJ2 HSJ2 HSPF4</td>\n",
       "      <td>397</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>44,868</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>P01911</td>\n",
       "      <td>DRB_HUMAN</td>\n",
       "      <td>HLA class II histocompatibility antigen, DRB1 ...</td>\n",
       "      <td>HLA-DRB1</td>\n",
       "      <td>266</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>29,966</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>DISULFID 44..108;  /evidence=\"ECO:0000269|PubM...</td>\n",
       "      <td>CARBOHYD 48;  /note=\"N-linked (GlcNAc...) aspa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>P01889</td>\n",
       "      <td>HLAB_HUMAN</td>\n",
       "      <td>HLA class I histocompatibility antigen, B alph...</td>\n",
       "      <td>HLA-B HLAB</td>\n",
       "      <td>362</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>40,460</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>DISULFID 125..188;  /evidence=\"ECO:0000255|PRO...</td>\n",
       "      <td>CARBOHYD 110;  /note=\"N-linked (GlcNAc...) asp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>P08246</td>\n",
       "      <td>ELNE_HUMAN</td>\n",
       "      <td>Neutrophil elastase (EC 3.4.21.37) (Bone marro...</td>\n",
       "      <td>ELANE ELA2</td>\n",
       "      <td>267</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>28,518</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>DISULFID 55..71;  /evidence=\"ECO:0000255|PROSI...</td>\n",
       "      <td>CARBOHYD 88;  /note=\"N-linked (GlcNAc...) aspa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>49795</td>\n",
       "      <td>Q9NZH0</td>\n",
       "      <td>GPC5B_HUMAN</td>\n",
       "      <td>G-protein coupled receptor family C group 5 me...</td>\n",
       "      <td>GPRC5B RAIG2</td>\n",
       "      <td>403</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>44,795</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CARBOHYD 30;  /note=\"N-linked (GlcNAc...) aspa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>49796</td>\n",
       "      <td>Q9NXW2</td>\n",
       "      <td>DJB12_HUMAN</td>\n",
       "      <td>DnaJ homolog subfamily B member 12</td>\n",
       "      <td>DNAJB12</td>\n",
       "      <td>375</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>41,860</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>49798</td>\n",
       "      <td>O95395</td>\n",
       "      <td>GCNT3_HUMAN</td>\n",
       "      <td>Beta-1,3-galactosyl-O-glycosyl-glycoprotein be...</td>\n",
       "      <td>GCNT3</td>\n",
       "      <td>438</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>50,864</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>DISULFID 70..227;  /evidence=\"ECO:0000250\"; DI...</td>\n",
       "      <td>CARBOHYD 289;  /note=\"N-linked (GlcNAc...) asp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>49803</td>\n",
       "      <td>Q9Y238</td>\n",
       "      <td>DLEC1_HUMAN</td>\n",
       "      <td>Deleted in lung and esophageal cancer protein ...</td>\n",
       "      <td>DLEC1 DLC1</td>\n",
       "      <td>1755</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>195,684</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>49805</td>\n",
       "      <td>Q8TDM6</td>\n",
       "      <td>DLG5_HUMAN</td>\n",
       "      <td>Disks large homolog 5 (Discs large protein P-d...</td>\n",
       "      <td>DLG5 KIAA0583 PDLG</td>\n",
       "      <td>1919</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>213,868</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20369 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Entry   Entry name                                      Protein names  \\\n",
       "0      P04439   HLAA_HUMAN  HLA class I histocompatibility antigen, A alph...   \n",
       "1      P31689  DNJA1_HUMAN  DnaJ homolog subfamily A member 1 (DnaJ protei...   \n",
       "2      P01911    DRB_HUMAN  HLA class II histocompatibility antigen, DRB1 ...   \n",
       "3      P01889   HLAB_HUMAN  HLA class I histocompatibility antigen, B alph...   \n",
       "4      P08246   ELNE_HUMAN  Neutrophil elastase (EC 3.4.21.37) (Bone marro...   \n",
       "...       ...          ...                                                ...   \n",
       "49795  Q9NZH0  GPC5B_HUMAN  G-protein coupled receptor family C group 5 me...   \n",
       "49796  Q9NXW2  DJB12_HUMAN                 DnaJ homolog subfamily B member 12   \n",
       "49798  O95395  GCNT3_HUMAN  Beta-1,3-galactosyl-O-glycosyl-glycoprotein be...   \n",
       "49803  Q9Y238  DLEC1_HUMAN  Deleted in lung and esophageal cancer protein ...   \n",
       "49805  Q8TDM6   DLG5_HUMAN  Disks large homolog 5 (Discs large protein P-d...   \n",
       "\n",
       "                         Gene names  Length              Organism     Mass  \\\n",
       "0                        HLA-A HLAA     365  Homo sapiens (Human)   40,841   \n",
       "1      DNAJA1 DNAJ2 HDJ2 HSJ2 HSPF4     397  Homo sapiens (Human)   44,868   \n",
       "2                          HLA-DRB1     266  Homo sapiens (Human)   29,966   \n",
       "3                        HLA-B HLAB     362  Homo sapiens (Human)   40,460   \n",
       "4                        ELANE ELA2     267  Homo sapiens (Human)   28,518   \n",
       "...                             ...     ...                   ...      ...   \n",
       "49795                  GPRC5B RAIG2     403  Homo sapiens (Human)   44,795   \n",
       "49796                       DNAJB12     375  Homo sapiens (Human)   41,860   \n",
       "49798                         GCNT3     438  Homo sapiens (Human)   50,864   \n",
       "49803                    DLEC1 DLC1    1755  Homo sapiens (Human)  195,684   \n",
       "49805            DLG5 KIAA0583 PDLG    1919  Homo sapiens (Human)  213,868   \n",
       "\n",
       "         Status                                     Disulfide bond  \\\n",
       "0      reviewed  DISULFID 125..188;  /evidence=\"ECO:0000255|PRO...   \n",
       "1      reviewed                                                NaN   \n",
       "2      reviewed  DISULFID 44..108;  /evidence=\"ECO:0000269|PubM...   \n",
       "3      reviewed  DISULFID 125..188;  /evidence=\"ECO:0000255|PRO...   \n",
       "4      reviewed  DISULFID 55..71;  /evidence=\"ECO:0000255|PROSI...   \n",
       "...         ...                                                ...   \n",
       "49795  reviewed                                                NaN   \n",
       "49796  reviewed                                                NaN   \n",
       "49798  reviewed  DISULFID 70..227;  /evidence=\"ECO:0000250\"; DI...   \n",
       "49803  reviewed                                                NaN   \n",
       "49805  reviewed                                                NaN   \n",
       "\n",
       "                                           Glycosylation  \n",
       "0      CARBOHYD 110;  /note=\"N-linked (GlcNAc...) asp...  \n",
       "1                                                    NaN  \n",
       "2      CARBOHYD 48;  /note=\"N-linked (GlcNAc...) aspa...  \n",
       "3      CARBOHYD 110;  /note=\"N-linked (GlcNAc...) asp...  \n",
       "4      CARBOHYD 88;  /note=\"N-linked (GlcNAc...) aspa...  \n",
       "...                                                  ...  \n",
       "49795  CARBOHYD 30;  /note=\"N-linked (GlcNAc...) aspa...  \n",
       "49796                                                NaN  \n",
       "49798  CARBOHYD 289;  /note=\"N-linked (GlcNAc...) asp...  \n",
       "49803                                                NaN  \n",
       "49805                                                NaN  \n",
       "\n",
       "[20369 rows x 10 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
